cpp/daal/include/services/internal/sycl/math/mkl_lapack.h

```diff
@@ -106,50 +125,72 @@ struct MKLPotrs
+private:
+    Status operator()(const math::UpLo uplo, const size_t n, const size_t ny, Buffer<algorithmFPType> & a, const size_t lda,
+                      Buffer<algorithmFPType> & b, const size_t ldb, const std::int64_t scratchpadSize)
+    {
+        using namespace daal::services;
 
-        cl::sycl::buffer<algorithmFPType, 1> b_sycl_buff = b.toSycl(status);
-        DAAL_CHECK_STATUS_VAR(status);
+        services::Status status;
+        const ::oneapi::fpk::uplo uplomkl = uplo == math::UpLo::Upper ? ::oneapi::fpk::uplo::upper : ::oneapi::fpk::uplo::lower;
 
+#ifdef DAAL_SYCL_INTERFACE_USM
+        if (a.isUSMBacked())
         {
-            const size_t minimalScratchpadSize =
-                size_t(::oneapi::fpk::lapack::potrs_scratchpad_size<algorithmFPType>(_queue, uplomkl, n, ny, lda, ldb));
-            if (scratchpad.get_count() < minimalScratchpadSize) return ErrorMemoryAllocationFailed;
+            auto a_usm = a.toUSM(status);
+            DAAL_CHECK_STATUS_VAR(status);
+
+            auto b_usm = b.toUSM(status);
+            DAAL_CHECK_STATUS_VAR(status);
+
+            algorithmFPType * scratchpad = nullptr;
+            if (scratchpadSize > 0)
+            {
+                scratchpad = cl::sycl::malloc<algorithmFPType>(scratchpadSize, _queue, cl::sycl::usm::alloc::shared);
+                if (scratchpad == nullptr) return ErrorMemoryAllocationFailed;
+            }
+
+            status |= catchSyclExceptions([&]() mutable {
+                ::oneapi::fpk::lapack::potrs(_queue, uplomkl, n, ny, a_usm.get(), lda, b_usm.get(), ldb, scratchpad, scratchpadSize);
+                _queue.wait_and_throw();
+            });
+
+            if (scratchpadSize > 0) cl::sycl::free(scratchpad, _queue);
+
+            scratchpad = nullptr;
         }
+        else
+#endif
+        {
+            cl::sycl::buffer<algorithmFPType, 1> a_sycl_buff = a.toSycl(status);
+            DAAL_CHECK_STATUS_VAR(status);
 
-        status |= catchSyclExceptions([&]() mutable {
-            ::oneapi::fpk::lapack::potrs(_queue, uplomkl, n, ny, a_sycl_buff, lda, b_sycl_buff, ldb, scratchpad, scratchpad.get_count());
-            _queue.wait_and_throw();
-        });
+            cl::sycl::buffer<algorithmFPType, 1> b_sycl_buff = b.toSycl(status);
+            DAAL_CHECK_STATUS_VAR(status);
 
-        return status;
-    }
+            status |= catchSyclExceptions([&]() mutable {
+                cl::sycl::buffer<algorithmFPType, 1> scratchpad { cl::sycl::range<1>(scratchpadSize) };
 
-    Status operator()(const math::UpLo uplo, const size_t n, const size_t ny, Buffer<algorithmFPType> & a, const size_t lda,
-                      Buffer<algorithmFPType> & b, const size_t ldb, const std::int64_t scratchpadSize)
-    {
-        Status status;
+                const size_t minimalScratchpadSize =
+                    size_t(::oneapi::fpk::lapack::potrs_scratchpad_size<algorithmFPType>(_queue, uplomkl, n, ny, lda, ldb));
```

Comment:
I think that there is lack of check for `minimalScratchpadSize`. Can it be negative?